codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+package io.shiftleft.utils
+
+import com.sun.management.OperatingSystemMXBean
+import org.slf4j.{Logger, LoggerFactory}
+
+import java.io.PrintWriter
+import java.lang.management.ManagementFactory
+import java.util
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.{Calendar, Date}
+import scala.collection.mutable
+case class TimeMetricRecordConfig(resultFile: String = "cpuandmemoryusage.csv", recordFreq: Int = 60000)
+case class MaxPrintBuffers(MAX_FULLNAME_BUFFER: Int = 90, MAX_NAME_BUFFER: Int = 55, MAX_DONE_TIME_BUFFER: Int = 40)
+object TimeMetric {
+  val cal = Calendar.getInstance()
+  // Stage wise performance will be only recorded if parallel recording is started through initialize()
+  val stagePerformance: mutable.Map[String, NumberProcessor] = mutable.Map()
+  private val start                                          = cal.getTime
+  private var subStageStartTime                              = start
+  private var newTime                                        = start
+  private var stageStartTime                                 = start
+  private val ACROSS_PROCESS_STAGE                           = "<ACROSS_ALL>"
+  private val STAGE_NOT_SET                                  = "<not set>"
+  private val STARTED                                        = "Started"
+  private val DONE                                           = "Done"
+  private var lastStageName                                  = STAGE_NOT_SET
+  private var lastSubStageName                               = STAGE_NOT_SET
+  private var additionalMetaData                             = STAGE_NOT_SET
+  private var alsoLogConsole                                 = false
+  private val logger: (String, Boolean) => Unit = (msg, subStage) => {
+    if (alsoLogConsole && (!subStage || !supressSubstages))
+      println(msg)
+    baseLogger.debug(msg)
+  }
+  private val baseLogger: Logger             = LoggerFactory.getLogger(getClass)
+  private var recorder: Option[TimeRecorder] = None
+  private var fullName: Boolean              = false
+  private var maxBuffers: MaxPrintBuffers    = MaxPrintBuffers()
+  private var supressSubstages: Boolean      = true
+  private lazy val maxStartingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    }
+  }
+  private lazy val maxEndingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER
+    }
+  }
+  private val osBean = ManagementFactory.getOperatingSystemMXBean().asInstanceOf[OperatingSystemMXBean]
+
+  def initialize(
```

Comment:
If we make `TimeMetric` a class, this code could be in the constructor (which I suppose is simply in the class body)

Reply:
The reason I made it SingleTon as this might needs to be accessed from Joern as well as from the projects extending the Joern. Creating a class will make it difficult to pass on that instance, as that will call for interface change in many places.  
That's what I feel. Singleton Object would be the best option in this scenario.

Reply:
@mpollmeier What do you think here?

Reply:
responded in a comment below: https://github.com/ShiftLeftSecurity/codepropertygraph/pull/1759#issuecomment-1991185785